src/corelibs/U2View/src/ov_sequence/ADVGraphModel.cpp

```diff
@@ -197,622 +138,305 @@ void GSequenceGraphDrawer::sl_calculationTaskFinished() {
-    QColor color;
-
-    if (lineColors.contains(graph->graphName)) {
-        color = lineColors.value(graph->graphName);
-    } else {
-        color = lineColors.value(DEFAULT_COLOR);
+    if (isUndefined(visibleMin) || isUndefined(visibleMax) || graphs.isEmpty() || graphs[0]->viewPoints.isEmpty()) {
+        p.drawText(rect, Qt::AlignCenter, tr("Graph is not available. Try to change calculation settings."));
+        return;
     }
 
-    foreach (GraphLabel *label, graph->graphLabels.getLabels()) {
-        label->setColor(color, color);
+    for (const QSharedPointer<GSequenceGraphData> &graph : qAsConst(graphs)) {
+        drawGraph(p, graph, rect);
+        const QList<GraphLabel *> &labels = graph->labels.getLabels();
+        for (GraphLabel *label : qAsConst(labels)) {
+            bool isVisible = updateLabel(graph, label, rect);
+            label->setVisible(isVisible);
+        }
     }
-
-    graph->graphLabels.getMovingLabel().setColor(color, Qt::red);
-}
-
-void GSequenceGraphDrawer::drawGraph(QPainter &p, const QSharedPointer<GSequenceGraphData> &d, const QRect &rect) {
-    float min = 0;
-    float max = 0;
-    PairVector points;
-    int nPoints = rect.width();
-    calculatePoints(d, points, min, max, nPoints);
-
-    if (!points.isEmpty()) {
-        assert(points.firstPoints.size() == nPoints);
-
-        double comin = commdata.minEdge, comax = commdata.maxEdge;
-        if (commdata.enableCuttoff) {
-            min = comin;
-            max = comax;
+    updateMovingLabels(graphs, rect);
+
+    // Draw min & max lines & labels.
+    QPen minMaxPen(Qt::DashDotDotLine);
+    minMaxPen.setWidth(1);
+    p.setPen(minMaxPen);
+    p.setFont(*defFont);
+
+    // Visible max value.
+    p.drawLine(rect.topLeft(), rect.topRight());
+    QRect maxTextRect(rect.x(), rect.y(), rect.width(), 12);
+    p.drawText(maxTextRect, Qt::AlignRight, QString::number(visibleMax, 'g', 4));
+
+    // Visible min value.
+    p.drawLine(rect.bottomLeft(), rect.bottomRight());
+    QRect minTextRect(rect.x(), rect.bottom() - 12, rect.width(), 12);
+    p.drawText(minTextRect, Qt::AlignRight, QString::number(visibleMin, 'g', 4));
+}
+
+void GSequenceGraphDrawer::drawGraph(QPainter &p, const QSharedPointer<GSequenceGraphData> &graph, const QRect &rect) const {
+    int width = rect.width();
+    SAFE_POINT(graph->viewPoints.size() == width, "Invalid count of view points in graph", );
+
+    QPen graphPen(Qt::SolidLine);
+    graphPen.setColor(seriesColorByName.value(seriesColorByName.contains(graph->graphName) ? graph->graphName : DEFAULT_COLOR));
+    graphPen.setWidth(1);
+    p.setPen(graphPen);
+
+    int graphHeight = rect.height() - 2;
+    float valueToHeightScale = visibleMin == visibleMax ? 1 : (graphHeight - 1) / (visibleMax - visibleMin);
+
+    int prevX = INT_MAX;  // INT_MAX is a default value that means that prevX/Y is not defined yet.
```

Comment:
So is that the maximum `int` value? If yes, it's better to use `numeric_limits`

Reply:
Could you please share more details why to prefer one over another?

Reply:
For the same reason we prefer to use `static constexpr` where it possible instead of `static const` - this is the compile-time computed variable (which means performance improving in particular and the most modern approach in general).

Reply:
Which of these 2 values is static const?